### PR TITLE
Fix fetching temporal column id from Raptor metadata

### PR DESCRIPTION
**Pull request by nileema**
Due to the following JDBI issue: https://github.com/jdbi/jdbi/issues/154
    we get a 0 as the column_id when the temporal_column_id is set to NULL.
    This fix works around that issue.

### CODE REVIEW

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardCompactionManager.java [hunk 1]
```diff
@@ -175,8 +176,8 @@ private void discoverShards()
                 continue;
             }
 
-            Long temporalColumnId = metadataDao.getTemporalColumnId(tableId);
-            if (temporalColumnId != null) {
+            OptionalLong temporalColumnId = metadataDao.getTemporalColumnId(tableId).getTemporalColumnId();
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardCompactionManager.java]: The SQL object method will return null if there are no rows (the `ResultSetMapper` is never called because there are no rows to map).

**Comment by electrum** [comment]
Can you explain what this is fixing?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/TemporalColumn.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+import org.skife.jdbi.v2.StatementContext;
+import org.skife.jdbi.v2.tweak.ResultSetMapper;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Objects;
+import java.util.OptionalLong;
+
+public class TemporalColumn
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/TemporalColumn.java]: We could have a more generic `OptionalLongMapper`. Though you still have to deal with null (in addition to not present).

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java [hunk 3]
```diff
@@ -213,6 +214,15 @@ int dropView(
 
     @SqlQuery("SELECT temporal_column_id\n" +
             "FROM tables\n" +
-            "WHERE table_id = :tableId")
+            "WHERE table_id = :tableId\n" +
+            "AND temporal_column_id IS NOT NULL")
     Long getTemporalColumnId(@Bind("tableId") long tableId);
+
+    @VisibleForTesting
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java]: I'm not sure this annotation is necessary. It seems like something the DAO should support anyway (and we might use in the future).

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java [hunk 4]
```diff
@@ -213,6 +214,15 @@ int dropView(
 
     @SqlQuery("SELECT temporal_column_id\n" +
             "FROM tables\n" +
-            "WHERE table_id = :tableId")
+            "WHERE table_id = :tableId\n" +
+            "AND temporal_column_id IS NOT NULL")
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java]: Indent `  AND`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java [hunk 5]
```diff
@@ -213,6 +214,15 @@ int dropView(
 
     @SqlQuery("SELECT temporal_column_id\n" +
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java]: Add a comment here about why the `IS NOT NULL` is needed

File: presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        H2EmbeddedDataSourceConfig dataSourceConfig = new H2EmbeddedDataSourceConfig().setFilename("mem:");
+        DataSource dataSource = new H2EmbeddedDataSource(dataSourceConfig);
+        DBI h2Dbi = new DBI(dataSource);
+        handle = h2Dbi.open();
+        dao = handle.attach(MetadataDao.class);
+        createMetadataTablesWithRetry(dao);
+    }
+
+    @AfterMethod
```
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java]: Add `alwaysRun = true`

File: presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import javax.sql.DataSource;
+
+import static com.facebook.presto.raptor.metadata.MetadataDaoUtils.createMetadataTablesWithRetry;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertNull;
+
+public class TestMetadataDao
```
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java]: Needs single threaded annotation

File: presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    private MetadataDao dao;
+    private Handle handle;
+
+    @BeforeMethod
+    public void setup()
+            throws Exception
+    {
+        H2EmbeddedDataSourceConfig dataSourceConfig = new H2EmbeddedDataSourceConfig().setFilename("mem:");
+        DataSource dataSource = new H2EmbeddedDataSource(dataSourceConfig);
+        DBI h2Dbi = new DBI(dataSource);
```
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java]: Inline this

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java [hunk 9]
```diff
@@ -213,6 +214,15 @@ int dropView(
...
-            "WHERE table_id = :tableId")
+            "WHERE table_id = :tableId\n" +
+            "AND temporal_column_id IS NOT NULL")
     Long getTemporalColumnId(@Bind("tableId") long tableId);
+
+    @VisibleForTesting
+    @SqlUpdate("UPDATE tables SET\n" +
+            "temporal_column_id = :columnId\n" +
+            "WHERE table_id = :tableId")
+    void updateTemporalColumnId(
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java]: How about `setTemporalColumnId`

File: presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Test
+    public void testTemporalColumn()
+            throws Exception
+    {
+        long tableId = dao.insertTable("default", "default", "table1");
+        dao.insertColumn(tableId, 1L, "col1", 1, "bigint");
+        dao.updateTemporalColumnId(1L, tableId);
+        Long columnId = dao.getTemporalColumnId(tableId);
+        assertNotNull(columnId);
```
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java]: Verify this has the expected value

File: presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        handle.close();
+    }
+
+    @Test
+    public void testTemporalColumn()
+            throws Exception
+    {
+        long tableId = dao.insertTable("default", "default", "table1");
+        dao.insertColumn(tableId, 1L, "col1", 1, "bigint");
+        dao.updateTemporalColumnId(1L, tableId);
```
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestMetadataDao.java]: Add a test between the insert and update that verifies it returns null

**Comment by electrum** [comment]
Looks good
